core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -80,6 +80,8 @@ abstract class MergingSnapshotProducer<ThisT> extends SnapshotProducer<ThisT> {
 
   // update data
   private final List<DataFile> newDataFiles = Lists.newArrayList();
+  private final CharSequenceSet newDataFilePaths = CharSequenceSet.empty();
+  private final CharSequenceSet newDeleteFilePaths = CharSequenceSet.empty();
```

Comment:
subclasses of `DataFile` / `DeleteFile` don't implement `equals()` / `hashCode()`, so using a separate set to store the file path